runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -3,7 +3,401 @@ pub(crate) mod error;
+    let mut account = source.program_account.clone();
+    account.set_data_from_slice(&data);
+    Ok(account)
+}
+
+impl Bank {
+    pub(crate) fn migrate_builtin_to_core_bpf(
+        &mut self,
+        builtin_program_id: &Pubkey,
+        config: &CoreBpfMigrationConfig,
+    ) -> Result<(), CoreBpfMigrationError> {
+        datapoint_info!(config.datapoint_name, ("slot", self.slot, i64));
+
+        let target =
+            TargetBuiltin::new_checked(self, builtin_program_id, &config.migration_target)?;
+        let source = SourceUpgradeableBpf::new_checked(self, &config.source_program_id)?;
+
+        // Attempt serialization first before modifying the bank.
+        let new_target_program_account = new_target_program_account(&target, &source)?;
+
+        // Gather old and new account data sizes, for updating the bank's
+        // accounts data size delta off-chain.
+        // The old data size is the total size of all accounts involved.
+        // The new data size is the total size of the source program accounts,
+        // since the target program account is replaced with a new program
+        // account of the same size as the source program account, and the
+        // source program data account is copied to the target program data
+        // account before both source program accounts are cleared.
+        let target_program_len = target.program_account.data().len();
+        let source_program_len = source.program_account.data().len();
+        let source_program_data_len = source.program_data_account.data().len();
+        let old_data_size = checked_add(
+            target_program_len,
+            checked_add(source_program_len, source_program_data_len)?,
+        )?;
+        let new_data_size = checked_add(source_program_len, source_program_data_len)?;
+
+        // Burn lamports from the target program account, since it will be
+        // replaced.
+        self.capitalization
+            .fetch_sub(target.program_account.lamports(), Relaxed);
+
+        // Replace the target builtin account with the
+        // `new_target_program_account` and clear the source program account.
+        self.store_account(&target.program_address, &new_target_program_account);
+        self.store_account(&source.program_address, &AccountSharedData::default());
+
+        // Copy the source program data account into the account at the target
+        // builtin program's data address, which was verified to be empty by
+        // `TargetBuiltin::new_checked`, then clear the source program data
+        // account.
+        self.store_account(&target.program_data_address, &source.program_data_account);
+        self.store_account(&source.program_data_address, &AccountSharedData::default());
+
+        // Remove the built-in program from the bank's list of built-ins.
+        self.builtin_program_ids.remove(&target.program_address);
+
+        // Unload the programs from the bank's cache.
```

Comment:
I suspect it is so that the target gets reloaded the next time it is used, right?
But it does not work that way because of forks. You have to insert a valid entry on deployment, not remove outdated entries. Otherwise the another fork might reload the program in the previous state and then you fork would conclude that that is what it should use too.

Reply:
@Lichtso maybe something like this?

```rust
// Update the bank's program cache.
let recompiled = self.load_program(&target.program_address, false, self.epoch());
let mut program_cache = self.transaction_processor.program_cache.write().unwrap();
program_cache.remove_programs([target.program_address, source.program_address].into_iter());
program_cache.assign_program(target.program_address, recompiled);
program_cache.assign_program(
    source.program_address,
    LoadedProgram::new_tombstone(self.slot(), LoadedProgramType::Closed),
);
```

It looks like the cache is only used in `load_program_with_pubkey` to fetch the current environment, so it doesn't seem unreasonable for that recompile step to come first. Plus, we do something similar in `_new_from_parent`:

https://github.com/anza-xyz/agave/blob/182d27f718e1d6df347bdbe60edaba6026f1c6b5/runtime/src/bank.rs#L1402-L1411

Reply:
`ProgramCache::remove_programs` should never be used (we should probably just delete it the method). Instead the cache will prune outdated entries when they become unreachable.

I think it would be better to just invoke the Loader-v3 deploy instruction here. That also makes sure we don't break the core program transition in the future.

Reply:
Ok I've added two new commits to do exactly as you've suggested, and driving the loader according to how we discussed offline.